data_structures/src/mainnet_validations.rs

```diff
@@ -16,6 +18,169 @@ pub const FIRST_HARD_FORK: Epoch = 192000;
 /// 28 April 2021 @ 9:00:00 UTC
 pub const SECOND_HARD_FORK: Epoch = 376320;
 
+/// TAPI Engine
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct TapiEngine {
+    /// bit votes counter by bits
+    pub bit_tapi_counter: BitTapiCounter,
+    /// wip activation
+    pub wip_activation: HashMap<String, Epoch>,
+}
+
+impl TapiEngine {
+    pub fn update_bit_counter(
+        &mut self,
+        v: u32,
+        epoch_to_update: Epoch,
+        block_epoch: Epoch,
+        avoid_wip_list: &HashSet<String>,
+    ) {
+        // In case of empty epochs, they would be considered as blocks with tapi version to 0
+        // In order to not update bit counter from old blocks where the block version was not used,
+        // the first time (bit_tapi_counter.last_epoch == 0) would be skipped in this conditional branch
+        if self.bit_tapi_counter.last_epoch != 0
+            && epoch_to_update > self.bit_tapi_counter.last_epoch + 1
```

Comment:
I'm not understanding or can't see the need for this guard.

Reply:
First time that you update witnet to have a TapiEngine, the last_epoch field starts to 0. It combined with the empty epoch checker would check the version of all of blocks of the chain, when it is not necessary. The idea is that when you start your node you set the current block epoch, or the first wip activation epoch in the "last_epoch" field without to recheck the complete chain

Reply:
Thanks for the clarification!